rclpy/rclpy/impl/logger_rcutils.py

```diff
@@ -0,0 +1,275 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from collections import OrderedDict
+try:
+    import cPickle as pickle
+except ImportError:
+    import pickle
+import importlib
+import inspect
+import time
+
+_rclpy_logging = importlib.import_module('._rclpy_logging', package='rclpy')
+
+_internal_callers = None  # List of known filenames from which logging methods can be called.
+
+
+def _find_caller(frame):
+    """Get the first calling frame that is outside of rclpy."""
+    global _internal_callers
+    if _internal_callers is None:
+        # Populate the list of known filenames from which logging methods can be called.
+        # This has to be done from within a function to avoid cyclic module imports.
+        import rclpy.logging
+        _internal_callers = [__file__, rclpy.logging.__file__]
```

Comment:
in 4d97de06660c8d0f2a4e3cd50bedf56cd7f765c2 I added another variable that third-parties can contribute to; I left the check on None for the internal filenames to avoid doing the importing + enforcing each time the function's called

Reply:
Imo having two variables makes it more complex than necessary. If you would rather not check if the items are already in the list you could use a global boolean flag to determine if `_find_caller` should add its own (on the first call) or not (on repeated calls).

Reply:
OK, I have change it to use a flag. (most of the change was in cdd4d2ada2a531a4a47226e83b6474efc1319be)